server.go

```diff
@@ -21,6 +22,51 @@ const (
 	SftpServerWorkerCount = 8
 )
 
+type FileLike interface {
+	Stat() (os.FileInfo, error)
+	ReadAt(b []byte, off int64) (int, error)
+	WriteAt(b []byte, off int64) (int, error)
+	Readdir(int) ([]os.FileInfo, error)
+	Name() string
+	Truncate(int64) error
+	Chmod(mode fs.FileMode) error
+	Chown(uid, gid int) error
+	Close() error
+}
+
+type dummyFile struct {
+}
+
+func (f *dummyFile) Stat() (os.FileInfo, error) {
+	return nil, os.ErrPermission
+}
+func (f *dummyFile) ReadAt(b []byte, off int64) (int, error) {
+	return 0, os.ErrPermission
+}
+func (f *dummyFile) WriteAt(b []byte, off int64) (int, error) {
+	return 0, os.ErrPermission
+}
+func (f *dummyFile) Readdir(int) ([]os.FileInfo, error) {
+	return nil, os.ErrPermission
+}
+func (f *dummyFile) Name() string {
+	return "dummyFile"
+}
+func (f *dummyFile) Truncate(int64) error {
+	return os.ErrPermission
+}
+func (f *dummyFile) Chmod(mode fs.FileMode) error {
+	return os.ErrPermission
+}
+func (f *dummyFile) Chown(uid, gid int) error {
+	return os.ErrPermission
+}
+func (f *dummyFile) Close() error {
+	return os.ErrPermission
+}
+
+var _ = dummyFile{} // ignore unused
```

Comment:
Hm, I think this however makes it super easy to extend the current codebase for more "virtual paths", for which only "a small subset of file operations" is implemented. One can now just "inherit" `dummyFile` (via promoted fields) and "overwrite" the methods one needs to implement.

> This also solves the problem you’ve created here where dummyFile.Name() always returns "dummyFile" which is a valid filename

Or one just removes "Name()" from the list of functions a default implementation is provided? What filename do you suggest?

Reply:
[YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

Reply:
How do you know? It doesn't add much complexity IMO. It allows one to do interesting stuff, for example one could create a /proc like "filesystem" that is completely implemented in the sftp server! I surely could see some interesting use cases for that, which I might even add later (in a private fork). I think one should consider the pros and cons, I don't see many cons in this case? But maybe I am missing something?

Reply:
If you want to add it to your private fork, I cannot stop you. But we don’t need this code in this package.

Reply:
Mixing `dummyFile` with `winRoot` IMO mixes two separate issues.. The first is to implement a "dummy" that realizes the `file` interface, the second is to implement the `winRoot` functionality.

When one looks at `winRoot`, one imediatly sees the gist of what it is doing. It is not obscurred by all the "dummy implementations", because those are contained in `dummyFile`. 

> But we don’t need this code in this package.

Who is "we"? Do you know who is going to fork it, and potentially want to add functionality? I don't think so.

> YAGNI

I think this principle is useful for deciding if one wants to invest the time into implementing something.. But in this case it is already implemented, and IMO is cleaner (for the reason described above).

If this gets into "No you have to change X and Y and Z because I like it more" without mentioning real reasons behind that, then I must tell you that I don't have time for that. It is similar to the "No you need to use `permission denied`, because I say so, and disregarding the point I gave on why I think `Operations cannot be performed on this handle` is more matching for a situation where not the privilege level of the user is the problem, but that the operation makes no sense for the referenced resource.

This is your repository so you are free to do what you want. But don't expect me to change this while not giving any clear reasons, thanks 🙏

Reply:
Thank you for the thorough explanation.

> Clients are far more likely to be operating under the assumption that os.ErrPermission is a “do not retry this operation, but handle is still open for further use”, while SSH_FX_FAILURE is so generic that a client cannot really recover from the failure programmatically at all, as it is completely undefined and unknown whether any further operations are possible at that point.

Okay, if that is the case I guess `SSH_FX_PERMISSION_DENIED` is practically the more appropriate choice.
However, I want to note, that it feels like `os.ErrPermission` then takes the role of "generic error" that SSH_FX_FAILURE perhaps should have.
And I still think that `os.ErrPermission` is / can be misleading to the end-user who reads this error message.

An error like `ENOSYS` would be ideal I guess. Hmm, `SSH_FX_OP_UNSUPPORTED` comes to mind.

> > indicates that an attempt was made to perform an operation which is not supported for the server
> [..]
> The server supports this operation. It just cannot perform this operation on that specific handle.

Well, one could say that the server does not implement e.g. copying a directory to `/` (the virtual root). But it _could_ be implemented to e.g. mean "create a new drive and copy the folder contents onto there".

So I think it would be more similar to when e.g. Linux gives `ENOSYS` (or `EOPNOTSUPP`) for deleting a file for a specific filesystem. It's not that linux is incapable of deleting files, but just the current implementation of the filesystem is (thus rendering linux unable to do that for that filesystem).

> The original form of Server has used a map[string]*os.File since it was first introduced in a merged in September 2015. We have only now 9 years later, seen good cause to convert this into an interface

If you think about it, this argument goes both ways. You originally said that implementing this would be a lot of work, and you don't see anyone having time for that. In effect, a useful feature thus might have never been implemented, because the code looks too "tedious" to be refactored/extended in such a way.

I see your point of keeping implementations "clean and lean". However, I'd say if it is not "a lot of complexity" that gets added, the advantages (easier to be extended in the future) outweighs the disadvantage.
And in this case, I personally even feel like separating `dummyFile` and `winRoot` is cleaner, even if one only uses `dummyFile` once.

But I guess it's a matter of taste, so I have changed the code according to your wishes.

> So, I must be patient, but insistent. If you interpret this professional courtesy as “passive aggressive”, then I’m sorry.

No, what I perceived as "passive aggressive" was when you just repeated things you already said, which I had given an answer to, but you (thereby) ignored what I said.
It felt as if you were not willing to listen to the points I made, and instead just were half-annoyed that I dared to question some of your statements.

Reply:
> Hmm, SSH_FX_OP_UNSUPPORTED comes to mind.

I had already addressed the appropriateness of this status value 5 days ago: https://github.com/pkg/sftp/issues/569#issuecomment-1879738231

> So I think it would be more similar to when e.g. Linux gives ENOSYS (or EOPNOTSUPP) for deleting a file for a specific filesystem.

Except that `ENOSYS` is not listed at all for `unlink(2)` https://man7.org/linux/man-pages/man2/unlink.2.html

In fact, instead, it is noted specifically that for linux:
```
       EPERM (Linux only)
              The filesystem does not allow unlinking of files.
```

`EOPNOTSUPP` is defined for _sockets only_, (https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_03)
```
[EOPNOTSUPP]
Operation not supported on socket. The type of socket (address family or protocol) does not support the requested operation. A conforming implementation may assign the same values for [EOPNOTSUPP] and [ENOTSUP].
```

You might note that it allows for it to be conflated with `ENOTSUP`. However, even when conflated, only socket operations should be returning the `EOPNOTSUPP` synonym, while other operations should still be returning `ENOTSUP`.

As for the appropriateness of `ENOTSUP` not only does my argument against `SSH_FX_OP_UNSUPPORTED` still apply, so does my argument against `ENOSYS`.

> because the code looks too "tedious" to be refactored/extended in such a way.

You’ve mistaken entirely what my anticipation of the tedium would be.

Your originally proposed method of implementing this was to modify every single point in the code where we called `getHandle` and treat a magic handle token specially. I warned you against that course of action, because this would have indeed been horrendously tedious. I directed you instead, to the least tedious implementation.

Instead, the tedium, which I anticipated, was never in constructing a pseudo-directory implementation. It was anticipating the process of ensuring everything worked properly, and that we covered all the “gotchas” that invariably result from implementing a parallel implementation—in this case to `*os.File`. Namely:
* if you call `Readdir(n)` with an `n > 0`, you have to ensure to return a most that number of entries.
* if you call `Readdir(n)` with an `n <= 0`, you have to ensure that it returns all entries.
* when you call `Readdir` a second time, you should not return any of the entries already returned.
* once `Readdir` is called and there are no remaining entries, it must return `io.EOF`, but not ever before
* the `Readdir` has to return filenames that actually reflect the usage that clients should be using them as (i.e. `c:` not just `c`)
* Checking for what `os.Stat("c:\")` actually returns, and how we need to change it in order to fit our use-case, so that we can be sure we’re returning the proper and appropriate values from all of the `fs.FileInfo` values for the individual drives
* Properly encoding the `toLocalPath` operations to ensure that we don’t allow accesses to `\Windows` but instead require a drive name

> No, what I perceived as "passive aggressive" was when you just repeated things you already said, which I had given an answer to, but you (thereby) ignored what I said.
> It felt as if you were not willing to listen to the points I made, and instead just were half-annoyed that I dared to question some of your statements.

If I have repeated myself, it is because you refused to acknowledge or listen to the point I had already made, or brought up a point we had already moved from, while trying to get me to relitigate it. (see, above, where you again drag out `SSH_FX_OP_UNSUPPORTED` after I had already written a detailed essay about why it would be inappropriate.)

Your arguments have repeatedly refused to actually listen to anything for which I was taking hours to research and compose, in order to patiently explain all this to you.

Even now, your response is not, “I understand. Thank you for taking so much time explaining these fine, minute, and subtle details, and helping to ensure that my code covers all of the gotcha corner cases involved in the standards. Let me know if this code addresses your concerns,” but instead you have answered, “I still refuse to acknowledge your arguments. I think I’m still right, but I have constructed a face-saving reason for me to comply. So, I will do it your way.”

You have mistaken “no matter how many times you repeat this argument, the standards and references continue to refute it,” for “I am ignoring your arguments.”

If I were ignoring your arguments, I would not be pulling up the POSIX documentation in order to evaluate the validity and soundness of your arguments. Meanwhile, in this single message alone, you’ve:
* repeated an already refuted `SSH_FX_OP_UNSUPPORTED`
* thrown out ENOSYS in a situation where under your very own scenario, linux is returning EPERM
* and tossed in `EOPNOTSUPP` without actually reading the POSIX standard to see that it is entirely constrained to sockets.

To me, it really does seems like you are not only not reading my explanations; you’re not even reading your own references.

----

Under advisement from @drakkan , this matter has gotten quite a bit out of hand, so we’re parking everything involved for 2 weeks, so that all parties (including me) can cool off, and we can come back to this later, when emotions are not being run raw.

Since I am both: the person with an imbalance of control in this situation, and also the one closing down the discussion; I want to be clear that I’m not doing this to silence you. You may post a reasonable response, but no further actions will be taken from my side for two weeks.†

†: With the caveat that there is one linting comment to address a typo, which will follow this.

Reply:
I feel like if we look at the bigger picture, we both mostly have the same goal of making this software more useful. It's not useful for either of us to spend much time discussing minor details, like whether to implement something in one way or a slightly different way.

> EOPNOTSUPP is defined for sockets only

Two things:

sshfs itself uses this error code when it gets an 'SSH_FX_OP_UNSUPPORTED':
https://github.com/libfuse/sshfs/blob/sshfs-3.7.3/sshfs.c#L1704

If we look at the linux source code, we see that it is actually used a lot in filesystem (fs/*) code:
https://elixir.bootlin.com/linux/v4.0/ident/EOPNOTSUPP

> As for the appropriateness of ENOTSUP not only does my argument against SSH_FX_OP_UNSUPPORTED still apply

Your argument was that the server does generally support the operation (e.g. unlink), so it should not return this error.
However, my counter-argument was that "operation" might entail the subject (e.g. the path) that is operated on.

If we look at the docs for `SSH_FX_PERMISSION_DENIED`, they say: `The user does not have sufficient permissions to perform the operation`.
So it seems that here the term "operation" clearly entails the subject (path) that is operated on, because the user probably can e.g. unlink his own files.

I don't think you have addressed that "counter-argument" yet, but instead you say things like `repeated an already refuted SSH_FX_OP_UNSUPPORTED`.

Maybe you have just missed/overlooked that, it's not (overall) a big deal. I think it will help us to assume that the other person had good intentions. Then probably it will be easier for us to see the point the other person is making, instead of going into some kind of "defensive mode".

> Instead, the tedium, which I anticipated, was never in constructing a pseudo-directory implementation

You said: `just even adding support to use a pseudo-directory in Server would be a pain, because right now our handles are all mapping directly to an actual *os.File which—as noted above—cannot exist for this pseudo-directory.`

This clearly references the restrictiveness of the old implementation, and that (alone) would be "a pain".

And true, making sure that everything is implemented correctly is quite some work too! And I am grateful that you took the time to review it that carefully.

> but instead you have answered “I still refuse to acknowledge your arguments. I think I’m still right, but I have constructed a face-saving reason for me to comply. So, I will do it your way.”

I said:
* Thank you for the thorough explanation.
* Okay, if that is the case I guess SSH_FX_PERMISSION_DENIED is practically the more appropriate choice
* I see your point of keeping implementations "clean and lean"

Still, I have my own opinions. And while I might do X as requested because I agree with you on Y, it doesn't mean that this has to "invalidate" my previous opinion Z which may also be about X.

> To me, it really does seems like you are not only not reading my explanations; you’re not even reading your own references.

I don't think such accusations are helpful in any way.

Besides, I am not even sure what you mean by "own references". I mostly just said that I think for the enduser a specific error code could be confusing. And that another one seems more fitting IMO.

I don't think that there is a clear "right" or "wrong", which is probably why one could argue about this endlessly. But I don't think there is much value in that discussion, as both "perspectives" are somewhat valid, and in the end its IMO a minor detail of this feature.